go/store/prolly/tree/json_location.go

```diff
@@ -0,0 +1,279 @@
+func jsonPathElementsFromMySQLJsonPath(pathBytes []byte) (path jsonLocation, err error) {
+	path = newRootLocation()
+	originalPathBytes := pathBytes
+	// TODO: We need to account for escaped characters
+	if pathBytes[0] != '$' {
+		return path, fmt.Errorf("invalid JSON key %s", originalPathBytes)
+	}
+	pathBytes = pathBytes[1:]
+	for len(pathBytes) > 0 {
+		var lastIndex int
+		if pathBytes[0] == '[' {
+			lastIndex = bytes.IndexByte(pathBytes, ']')
+			arrayIndex, err := strconv.Atoi(string(pathBytes[1:lastIndex]))
+			if err != nil {
+				return path, err
+			}
+			path.appendArrayIndex(uint64(arrayIndex))
+		} else if pathBytes[0] == '.' {
+			pathBytes = pathBytes[1:]
+			lastIndex = min(bytes.IndexByte(pathBytes, '['),
+				bytes.IndexByte(pathBytes, '.'))
+			if lastIndex == -1 {
+				path.appendObjectKey(pathBytes)
+				return path, err
+			}
+			path.appendObjectKey(pathBytes[:lastIndex])
+		} else {
+			return path, fmt.Errorf("invalid JSON key %s", originalPathBytes)
+		}
+		pathBytes = pathBytes[lastIndex:]
+	}
+	return path, nil
+}
+
+func (p *jsonLocation) appendObjectKey(key []byte) {
+	p.key = append(append(p.key, beginObjectKey), key...)
+	p.offsets = append(p.offsets, len(p.key))
+}
+
+func (p *jsonLocation) appendArrayIndex(idx uint64) {
+	p.key = append(append(p.key, beginArrayKey), getVarInt(idx)...)
+	p.offsets = append(p.offsets, len(p.key))
+}
+
+func (p *jsonLocation) appendEncodedArrayIndex(idx []byte) {
+	p.key = append(append(p.key, beginArrayKey), idx...)
+	p.offsets = append(p.offsets, len(p.key))
+}
+
+func (p *jsonLocation) pop() {
+	lastOffset := p.offsets[len(p.offsets)-2]
+	p.offsets = p.offsets[:len(p.offsets)-1]
+	p.key = p.key[:lastOffset]
+}
+
+func (p *jsonLocation) setScannerState(s scannerState) {
+	p.key[0] = byte(s)
+}
+
+func (p *jsonLocation) getScannerState() scannerState {
+	return scannerState(p.key[0])
+}
+
+func (p *jsonLocation) getPathElement(i int) (key []byte, isArray bool) {
+	start := p.offsets[i]
+	end := p.offsets[i+1]
+	key = p.key[start+1 : end]
+	isArray = p.key[start] == beginArrayKey
+	return key, isArray
+}
+
+func (p *jsonLocation) size() int {
+	return len(p.offsets) - 1
+}
+
+func (p *jsonLocation) getLastPathElement() (key []byte, isArray bool) {
+	return p.getPathElement(p.size() - 1)
+}
+
+func (p *jsonLocation) Clone() jsonLocation {
+	return jsonLocation{
+		key:     bytes.Clone(p.key),
+		offsets: slices.Clone(p.offsets),
+	}
+}
+
+// compareJsonLocations creates an ordering on locations by determining which one would come first in a normalized JSON
+// document where all keys are sorted lexographically.
+func compareJsonLocations(left, right jsonLocation) int {
+	minLength := min(left.size(), right.size())
+	for i := 0; i < minLength; i++ {
+		l, _ := left.getPathElement(i)
+		r, _ := right.getPathElement(i)
+		c := bytes.Compare(l, r)
```

Comment:
Not really, since it shouldn't ever be possible to compare one against the other; you can't have a value with both keys and indexes.

In practice, MySQL does have some corner cases around indexing into non-arrays, but I don't think here would ever be the place to detect those anyway.